### PR TITLE
Named functions

### PR DESCRIPTION
**Pull request by endJunction**
See commit msgs.

### CODE REVIEW

File: NumLib/NamedFunctionCaller.cpp [hunk 1]
```diff
@@ -10,44 +10,63 @@
 #include "NamedFunctionCaller.h"
 
 #include <algorithm>
-#include <list>
+#include <queue>
```
Comment (chleh) [inline NumLib/NamedFunctionCaller.cpp]: I think that's not necessary, is it?

File: NumLib/NamedFunctionCaller.cpp [hunk 2]
```diff
@@ -10,44 +10,63 @@
...
 
 #include "BaseLib/uniqueInsert.h"
 #include "BaseLib/Error.h"
 
-bool hasTopologicalOrdering(std::vector<std::vector<int>> const& dependencies)
+/// To understand the working of the algorithm the following two lemmata are
+/// useful:
+/// If a graph has a topological order, then it is a Directed Acyclic Graph
```
Comment (chleh) [inline NumLib/NamedFunctionCaller.cpp]: Pedantic: If a digraph... ^^

File: NumLib/NamedFunctionCaller.cpp [hunk 3]
```diff
@@ -10,44 +10,63 @@
...
+
     while (!q.empty())
     {
-        auto const fct_idx = q.front();
-        q.pop_front();
-        for (int dep : dependencies[fct_idx]) {
-            if (dep < 0) continue;
-            if (--dep_counts[dep] == 0) {
-                q.push_back(dep);
+        auto const node_i = q.front();
```
Comment (chleh) [inline NumLib/NamedFunctionCaller.cpp]: `q.back()` because you `pop_back()`

Comment (endJunction) [inline NumLib/NamedFunctionCaller.cpp]: Good catch, but the tests didn't fail... ;(

Comment (chleh) [inline NumLib/NamedFunctionCaller.cpp]: In order to make a test fail the queue must have at least two elements at some stage, e.g., if there are two nodes in the graph that have no incoming edges.

**Comment by chleh** [comment]
Thanks for your improvements and comments.

**Pull request merged**